hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/dao/SpanLogEventFetcher.java

```diff
@@ -0,0 +1,192 @@
+      ArgumentDeserializer argumentDeserializer,
+      FilterRequestBuilder filterRequestBuilder,
+      GatewayServiceFutureStubProvider gatewayServiceFutureStubProvider,
+      GraphQlGrpcContextBuilder grpcContextBuilder) {
+    this.attributeConverter = attributeConverter;
+    this.filterConverter = filterConverter;
+    this.attributeMapConverter = attributeMapConverter;
+    this.argumentDeserializer = argumentDeserializer;
+    this.filterRequestBuilder = filterRequestBuilder;
+    this.gatewayServiceStub = gatewayServiceFutureStubProvider.get();
+    this.grpcContextBuilder = grpcContextBuilder;
+  }
+
+  /**
+   *
+   *
+   * <ul>
+   *   <li>1. Fetch log event attributes from {@code gqlRequest}
+   *   <li>2. Build log event request using attribute and spanIds as filter
+   *   <li>3. Query log events
+   *   <li>4. Processed log events response to build mapping from spanId to logEvent
+   * </ul>
+   */
+  Single<SpanLogEventsResponse> fetchLogEvents(
+      SpanRequest<OrderArgument> gqlRequest, SpansResponse spansResponse) {
+    if (null == gqlRequest.spanEventsRequest().idAttribute()
+        || null == gqlRequest.logEventAttributes()
+        || gqlRequest.logEventAttributes().isEmpty()) {
+      return Single.just(new SpanLogEventsResponse(spansResponse, Map.of()));
+    }
+    return buildLogEventsRequest(gqlRequest, spansResponse)
+        .flatMap(
+            logEventsRequest ->
+                makeRequest(gqlRequest.spanEventsRequest().context(), logEventsRequest))
+        .flatMap(
+            logEventsResponse ->
+                buildResponse(gqlRequest.logEventAttributes(), spansResponse, logEventsResponse));
+  }
+
+  private Single<LogEventsRequest> buildLogEventsRequest(
+      SpanRequest<OrderArgument> gqlRequest, SpansResponse spansResponse) {
+    return zip(
+        this.attributeConverter.convert(gqlRequest.logEventAttributes()),
+        buildLogEventsQueryFilter(gqlRequest, spansResponse).flatMap(filterConverter::convert),
+        (selections, filter) ->
+            LogEventsRequest.newBuilder()
+                .setStartTimeMillis(
+                    gqlRequest.spanEventsRequest().timeRange().startTime().toEpochMilli())
+                .setEndTimeMillis(
+                    gqlRequest.spanEventsRequest().timeRange().endTime().toEpochMilli())
+                .addAllSelection(selections)
+                .setFilter(filter)
+                .build());
+  }
+
+  private Single<List<AttributeAssociation<FilterArgument>>> buildLogEventsQueryFilter(
+      SpanRequest<OrderArgument> gqlRequest, SpansResponse spansResponse) {
+    List<String> spanIds =
+        spansResponse.getSpansList().stream()
+            .map(
+                v ->
+                    v.getAttributesMap()
+                        .get(gqlRequest.spanEventsRequest().idAttribute().attribute().id())
+                        .getString())
+            .collect(Collectors.toList());
+    Map<String, Object> argMap =
```

Comment:
Rather than artificially pushing a map through the deserializer, it's usually easier to just create a new subtype of filter for local use. See an almost identical use case: `EntityIdFilter`